cupy/_core/_routines_indexing.pyx

```diff
@@ -262,15 +263,31 @@ cpdef list _prepare_slice_list(slices):
                 # keep scalar int
                 continue
 
-        if cupy.min_scalar_type(s).char == 'O':
-            raise IndexError(
-                'arrays used as indices must be of integer (or boolean) type')
         try:
             s = core.array(s, dtype=None, copy=None)
-        except ValueError:
-            # "Unsupported dtype"
+        except ValueError as e:
+            # Conversion failed, presumably with "Unsupported dtype".
+
+            # If this is a NumPy array, it should have an unsupportd dtype
+            # raise that as as not "integer or boolean" dtype.
+            if isinstance(s, numpy.ndarray):
```

Comment:
This is a special edge case, but since `s` is a tensor of another CUDA tensor library that supports the CUDA array interface and has a dtype not supported by CuPy, this line might need to evaluate to `True`.

Reply:
Hmmm good point, it's surprisingly tricky (to the point that I wouldn't bother making NumPy 100%)!

I'll let it sink in for a moment. `min_scalar_type` is probably pragmatic here in the end.
I think I'll apply this and that should get us as close as we reasonably can:
```
    try:
        dtype = min_scalar_type(s)
    except Exception:
        raise e from None  # raise original error

    if dtype.char in _dtype.all_type_chars:
        raise  # Conversion failed but dtype is fine, original error.
    elif dtype.char != "O" or isinstance(s, numpy.ndarray):
        raise IndexError('arrays used as indices...')
    
     # Object dtype but not a NumPy array is probably an arbitrary
     # object (that can be wrapped into a 0-D array).
     raise IndexError(generic_message)
```

Reply:
Indeed, as you say, this case is unlikely to become an issue. Let's revisit it if someone opens an issue.